drivers/retained_mem/retained_mem_zephyr_reg.c

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (c) 2023, Nordic Semiconductor ASA
+ * Copyright (c) 2023, Bjarki Arge Andreasen
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT zephyr_retained_reg
+
+#include <string.h>
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/retained_mem.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(retained_mem_zephyr_reg, CONFIG_RETAINED_MEM_LOG_LEVEL);
+
+#ifdef CONFIG_RETAINED_MEM_MUTEXES
+struct zephyr_retained_mem_reg_data {
```

Comment:
Would not be better to declare an empty structure and ifdefed the mutex member only? Then `DEVICE_DT_INST_DEFINE` macro would be less complicated.

Reply:
+1 to that

Reply:
I'm not sure that is possible, I don't think an empty C structure is allowed

Reply:
You are correct, the data pointer is just set to some random value if the struct is empty :)

Reply:
gotta love C :-)

Reply:
Generally, it's UB but as @fabiobaltieri said - it should work in C.